### PR TITLE
[ecs] Rework `Entity` to use `u64`

### LINKED ISSUE #2346: Entity cmp::Eq improvement
## What problem does this solve or what need does it fill?

Current Entity comparison relies on rust's `#derive(Eq, ...)`

It now actually makes two comparisons (see asm code)

https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=38662860685c22c6cbaf0a14a7e38b8b

## What solution would you like?

If treat whole Entity struct as u64 (through pointer cast on comparison, or even by store generation+id as u64) this could become one bitwise comparison.

https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=3351c4c203e76073de253b3301ed8d13

### PR DESCRIPTION
**Pull request by NathanSWard**
# Objective

- Currently `Entity` is defined as `struct Entity { id: u32, gen: u32 }` however, this produces sub-optimal assembly for various operations (including: `Eq`, `Ord`, `Copy`, and `new`)
- Fixes #2346

## Solution

- Have entity be defined as a single `Entity(u64)` and use some bit manipulation to extract/store the id/generation numbers. 
- This ends up producing better assembly in almost all cases.

### CODE REVIEW

**Comment by NathanSWard** [comment]
In case anyone is curious, here is a rust playground link to see the different assembly outputs.
https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=3ae1444da8f9d989e522b6234d836bb0

Most notably the `new` (or as I call `v2`) implementation has nicer assembly for:
- `new(id: u32)`
- `from_id_and_gen(id: u32, gen: u32)`
- `std::cmp::Eq`
- `std::cmp::Ord`
- `Copy`

the old and new have the same assembly for retrieving the `id`

the old version has one fewer instruction for retrieving the `generation`. (as the new one requires a `shrq $32, <reg>`

**Comment by mockersf** [comment]
could you run the benches from https://github.com/cart/ecs_bench_suite/tree/bevy-benches before/after (hopefully they don't need updating...)?

**Comment by TheRawMeatball** [comment]
Hmm, has anyone tried out unsafe optimizations yet? I tried adding `#[repr(align(8))]` to the old impl, and while it didn't improve all of them, it did have some positive effects. Also, I don't think testing with constants is very fair for the assembly output, as when they're changed to variables the new with id and gen function actually has one more instruction than the old impl.

**Comment by NathanSWard** [comment]
>  I don't think testing with constants is very fair for the assembly output, as when they're changed to variables the new with id and gen function actually has one more instruction than the old impl.

Yep, that's fair!
I updated the link/description to better reflect this.

**Comment by NathanSWard** [comment]
> Hmm, has anyone tried out unsafe optimizations yet?
I'm not quite sure about which optimization you're talking about, however in the original linked issues, there's the recommendation to define `Eq` as:
```rust
impl Eq for Entity {
    fn eq(&self, other: &Self) -> bool {
        *(self as *const Entity as *const u64) == *(other as *const Entity as *const u64)
    }
}
```
which outputs the better assembly.

If we went this path, we (should) do it for `Eq` and `Ord`.

(The one that still really confuses me is the `Copy` assembly....

**Comment by TheRawMeatball** [comment]
Hmm, it looks like using arrays gets us *almost* all the way there, but is arguably easier to understand than bit manipulation: https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=7c664f861a7d533cbe457cf37c7cb9a0

The only function that's different is the cmp impl, but this impl is never used in actual code, and more importantly could still be improved using unsafe: <https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=2fbce7c58019f4a98bb3d81f7e638515>

**Comment by bjorn3** [comment]
> (The one that still really confuses me is the Copy assembly....

The ABI is different. For `EntityOld` the abi is `ScalarPair` meaning that it is passed in two registers. For `EntityNew` the abi is `Scalar` meaning that it is passed in a single register. (You can annotate a type definition or alias with `#[rustc_layout(abi)]` on nightly to see the abi or with `#[rustc_layout(debug)]` to see the full layout.)

**Comment by NathanSWard** [comment]
> Hmm, it looks like using arrays gets us almost all the way there, but is arguably easier to understand than bit manipulation: 

I agree, I think the array is a little easier to understand, however, if we can avoid `unsafe` that's ideal (imo). 
So I would have a slight preference for `Entity(u64)` however, I'm also more than happy to use `Entity([u32;2])`

File: crates/bevy_ecs/src/entity/mod.rs [hunk 1]
```diff
@@ -21,15 +21,18 @@ use std::{
 /// Components of a specific entity can be accessed using
 /// [`Query::get`](crate::system::Query::get) and related methods.
 #[derive(Clone, Copy, Hash, Eq, Ord, PartialEq, PartialOrd)]
-pub struct Entity {
-    pub(crate) generation: u32,
-    pub(crate) id: u32,
-}
+pub struct Entity(u64);
```
Comment (DJMcNab) [inline crates/bevy_ecs/src/entity/mod.rs:24]: We should have a non-doc comment here saying what scheme we use and why

File: crates/bevy_ecs/src/entity/mod.rs [hunk 2]
```diff
@@ -21,15 +21,18 @@ use std::{
...
 
 impl Entity {
+    #[inline(always)]
+    fn from_id_and_gen(id: u32, gen: u32) -> Self {
+        Self(u64::from(gen) << 32 | u64::from(id))
+    }
+
     /// Creates a new entity reference with a generation of 0.
+    #[inline(always)]
     pub fn new(id: u32) -> Entity {
```
Comment (DJMcNab) [inline crates/bevy_ecs/src/entity/mod.rs:34]: Do we actually want this function?

**Review by DJMcNab — COMMENTED** [review]
impl looks good, I have no thoughts on the performance.

File: crates/bevy_ecs/src/entity/mod.rs (hunk 2 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/entity/mod.rs:34]: I've found this useful for writing tests before.

**Comment by NathanSWard** [comment]
Closing for now as we don't seem to have a strong argument for adding this extra complexity.

**Pull request closed without merging**